src/blockdata/script.rs

```diff
@@ -324,20 +796,122 @@ fn read_uint_iter(data: &mut core::slice::Iter<'_, u8>, size: usize) -> Result<u
+    // of `for`.
+    while let Some(byte) = iter.next() {
+        let opcode = opcodes::All::from(*byte);
+
+        let data_len = if let opcodes::Class::PushBytes(n) = opcode.classify(opcodes::ClassifyContext::Legacy) {
+            n as usize
+        } else {
+            match opcode {
+                opcodes::all::OP_PUSHDATA1 => {
+                    // side effects: may write and break from the loop
+                    read_push_data_len!(&mut iter, 1, f)
+                }
+                opcodes::all::OP_PUSHDATA2 => {
+                    // side effects: may write and break from the loop
+                    read_push_data_len!(&mut iter, 2, f)
+                }
+                opcodes::all::OP_PUSHDATA4 => {
+                    // side effects: may write and break from the loop
+                    read_push_data_len!(&mut iter, 4, f)
+                }
+                _ => 0
+            }
+        };
+
+        if at_least_one {
+            f.write_str(" ")?;
+        } else {
+            at_least_one = true;
+        }
+        // Write the opcode
+        if opcode == opcodes::all::OP_PUSHBYTES_0 {
+            f.write_str("OP_0")?;
+        } else {
+            write!(f, "{:?}", opcode)?;
+        }
+        // Write any pushdata
+        if data_len > 0 {
+            f.write_str(" ")?;
+            if data_len <= iter.len() {
+                for ch in iter.by_ref().take(data_len) {
+                    write!(f, "{:02x}", ch)?;
+                }
+            } else {
+                f.write_str("<push past end>")?;
+                break;
+            }
+        }
+    }
+    Ok(())
+}
+
+impl ScriptBuf {
     /// Creates a new empty script.
-    pub fn new() -> Script { Script(vec![].into_boxed_slice()) }
+    pub fn new() -> Self {
+        ScriptBuf(Vec::new())
+    }
+
+    /// Creates a new empty script with pre-allocated capacity.
+    pub fn with_capacity(capacity: usize) -> Self {
+        ScriptBuf(Vec::with_capacity(capacity))
+    }
+
+    /// Pre-allocates `additional_len` bytes if needed.
+    pub fn reserve(&mut self, additional_len: usize) {
+        self.0.reserve(additional_len);
+    }
+
+    /// Returns a reference to unsized script.
+    pub fn as_script(&self) -> &Script {
+        Script::from_bytes(&self.0)
+    }
+
+    /// Returns a mutable reference to unsized script.
+    pub fn as_mut_script(&mut self) -> &mut Script {
+        Script::from_bytes_mut(&mut self.0)
+    }
 
     /// Generates P2PK-type of scriptPubkey.
-    pub fn new_p2pk(pubkey: &PublicKey) -> Script {
+    pub fn new_p2pk(pubkey: &PublicKey) -> Self {
         Builder::new()
             .push_key(pubkey)
             .push_opcode(opcodes::all::OP_CHECKSIG)
             .into_script()
     }
 
     /// Generates P2PKH-type of scriptPubkey.
-    pub fn new_p2pkh(pubkey_hash: &PubkeyHash) -> Script {
+    pub fn new_p2pkh(pubkey_hash: &PubkeyHash) -> Self {
```

Comment:
With two types so similar in the same file perhaps its clearer to return `Script` instead of `Self`?

Reply:
I always use `Self` for return types to make refactorings less painful. Although ideally there won't be any such refactorings...